File: ports/go/security/renderer_divergence.go
package security

import "sync"

// renderer-divergence — detection of codepoint/sequence shapes known to render
// differently across font + terminal + browser stacks (the display-layer
// detector, layer D).
//
// Byte-faithful port of Unicode/Security/Display/RendererDivergence.lean,
// mirroring the verified Rust port
// security/display/renderer_divergence.rs.
//
// Threat model. An adversary crafts content that renders one way in the
// auditor's renderer (a benign glyph or an empty span) and a different way in
// the consumer's renderer (a misleading glyph, a wider glyph, or a different
// sequence). This is the "fingerprint stability" family — clear inputs render
// the same across the renderer cohort the Standard documents as stable.
//
// What the detector draws. A heuristic three-value split, surfaced through the
// universal clear/hazard carrier: an input is clear when none of the documented
// variance triggers fire, and otherwise is classified by the first trigger in
// priority order — combining-mark stack overflow, variation-selector presence,
// an unregistered ZWJ shape, fullwidth/halfwidth display, or mixed direction.
// It reuses the port's own tables (the variation-selector predicate
// isVariationSelector, the Grapheme_Extend property from the bundled
// DerivedCoreProperties.txt, the RGI ZWJ registry
// ezwjIsRegisteredZwjSequence, and the strong-bidi predicates
// isStrongLtr / isStrongRtl), never a host rendering or shaping library.
//
// Sub-threats (priority order):
//	1. CombiningStackOverflow    Zalgo-like combining-mark stack >= 4 on a base.
//	2. VariationSelectorVariance any variation selector present.
//	3. UnregisteredZwjVariance   ZWJ-containing input not in the RGI ZWJ set.
//	4. FullwidthVariance         a fullwidth/halfwidth codepoint present.
//	5. MixedDirectionVariance    both strong-LTR and strong-RTL codepoints.

// ─────────────────────────────────────────────────────────────────────
// §1 Constants
// ─────────────────────────────────────────────────────────────────────

// rdMinCombiningStack is the combining-mark stack depth (on a single base) at or
// beyond which the input is treated as a Zalgo-style rendering-variance hazard.
const rdMinCombiningStack = 4

// rdZWJ is the ZERO WIDTH JOINER codepoint.
const rdZWJ uint32 = 0x200D

// ─────────────────────────────────────────────────────────────────────
// §2 Types
// ─────────────────────────────────────────────────────────────────────

// rdSubThreat is the sub-threat this detector fired, in the port's
// struct-with-tag idiom (mirroring ezwjSubThreat). tag is the human-facing
// classification tag; the remaining fields carry the sub-threat's data (only
// those relevant to the fired tag are meaningful). The tag values, in priority
// order, are: CombiningStackOverflow, VariationSelectorVariance,
// UnregisteredZwjVariance, FullwidthVariance, MixedDirectionVariance.
type rdSubThreat struct {
	tag string
	// basePos — CombiningStackOverflow: position of the base the stack sits on.
	basePos int
	// stackLen — CombiningStackOverflow: the stack depth tested (>= rdMinCombiningStack).
	stackLen int
	// firstVsPos — VariationSelectorVariance: position of the first variation selector.
	firstVsPos int
	// firstVsCp — VariationSelectorVariance: codepoint of the first variation selector.
	firstVsCp uint32
	// firstZwjPos — UnregisteredZwjVariance: position of the first ZWJ.
	firstZwjPos int
	// firstFwPos — FullwidthVariance: position of the first fullwidth/halfwidth codepoint.
	firstFwPos int
	// firstFwCp — FullwidthVariance: the fullwidth/halfwidth codepoint.
	firstFwCp uint32
	// ltrCount — MixedDirectionVariance: count of strong-LTR codepoints.
	ltrCount int
	// rtlCount — MixedDirectionVariance: count of strong-RTL codepoints.
	rtlCount int
}

// rdClassification is the top-level RendererDivergence classification. When
// clear is true rendering is consistent across the documented renderer cohort;
// otherwise sub names the fired sub-threat, positions holds the implicated
// codepoint positions, and decoded is the decoded-byte projection (always empty
// here, kept for shape parity with the Lean Classification.hazard).
type rdClassification struct {
	clear     bool
	sub       rdSubThreat
	positions []int
	decoded   []uint8
}

// isClear reports whether the classification is Clear (i.e. stable).
func (c rdClassification) isClear() bool {
	return c.clear
}

// tag is the human-facing tag for a hazard; the second result is false when
// clear.
func (c rdClassification) tag() (string, bool) {
	if c.clear {
		return "", false
	}
	return c.sub.tag, true
}

// posns returns the implicated positions (empty when clear).
func (c rdClassification) posns() []int {
	if c.clear {
		return []int{}
	}
	return c.positions
}

// rdVerdict is the structured output of rendererDivergenceDetect (mirrors the
// Lean Verdict).
type rdVerdict struct {
	input          []uint32
	classify       rdClassification
	vsCount        int
	combiningCount int
	fullwidthCount int
	hasZwj         bool
	strongLtrCount int
	strongRtlCount int
}

// ─────────────────────────────────────────────────────────────────────
// §3 Grapheme_Extend data (reuses the bundled DerivedCoreProperties.txt)
// ─────────────────────────────────────────────────────────────────────

var (
	rdGraphemeExtendOnce   sync.Once
	rdGraphemeExtendRanges [][2]uint32
)

// rdGraphemeExtendProperty reports whether cp carries the Grapheme_Extend core
// property, read from the port's own bundled DerivedCoreProperties.txt via the
// shared parseCasingProperty machinery (no new data file, no host library).
func rdGraphemeExtendProperty(cp uint32) bool {
	rdGraphemeExtendOnce.Do(func() {
		rdGraphemeExtendRanges = parseCasingProperty("Grapheme_Extend")
	})
	return inRanges(rdGraphemeExtendRanges, cp)
}

// ─────────────────────────────────────────────────────────────────────
// §4 Core predicates
// ─────────────────────────────────────────────────────────────────────

// rdIsZwj reports whether cp is the ZWJ codepoint.
func rdIsZwj(cp uint32) bool {
	return cp == rdZWJ
}

// rdIsGraphemeExtend reports whether cp has Grapheme_Cluster_Break = Extend.
//
// The port bundles DerivedCoreProperties.txt (Grapheme_Extend) but not
// GraphemeBreakProperty.txt. The two sets are identical here save for one range:
// the emoji skin-tone modifiers U+1F3FB..U+1F3FF, which have GCB = Extend but
// are not Grapheme_Extend. The port already carries that range as
// ezwjIsEmojiModifier, so their union reproduces the verified Rust reference's
// grapheme::is_grapheme_extend byte-for-byte, using only tables already bundled.
func rdIsGraphemeExtend(cp uint32) bool {
	return rdGraphemeExtendProperty(cp) || ezwjIsEmojiModifier(cp)
}

// ─────────────────────────────────────────────────────────────────────
// §5 Sub-detectors
// ─────────────────────────────────────────────────────────────────────

func rdCountVs(input []uint32) int {
	n := 0
	for _, cp := range input {
		if isVariationSelector(cp) {
			n++
		}
	}
	return n
}

func rdCountCombining(input []uint32) int {
	n := 0
	for _, cp := range input {
		if rdIsGraphemeExtend(cp) {
			n++
		}
	}
	return n
}

func rdCountFullwidth(input []uint32) int {
	n := 0
	for _, cp := range input {
		if isFullwidthHalfwidth(cp) {
			n++
		}
	}
	return n
}

func rdInputHasZwj(input []uint32) bool {
	for _, cp := range input {
		if rdIsZwj(cp) {
			return true
		}
	}
	return false
}

func rdCountStrongLtr(input []uint32) int {
	n := 0
	for _, cp := range input {
		if isStrongLtr(cp) {
			n++
		}
	}
	return n
}

func rdCountStrongRtl(input []uint32) int {
	n := 0
	for _, cp := range input {
		if isStrongRtl(cp) {
			n++
		}
	}
	return n
}

// rdFirstVsPos returns the position and codepoint of the first variation selector.
func rdFirstVsPos(input []uint32) (int, uint32, bool) {
	for idx, cp := range input {
		if isVariationSelector(cp) {
			return idx, cp, true
		}
	}
	return 0, 0, false
}

// rdFirstZwjPos returns the position of the first ZWJ.
func rdFirstZwjPos(input []uint32) (int, bool) {
	for idx, cp := range input {
		if rdIsZwj(cp) {
			return idx, true
		}
	}
	return 0, false
}

// rdFirstFullwidthPos returns the position and codepoint of the first
// fullwidth/halfwidth codepoint.
func rdFirstFullwidthPos(input []uint32) (int, uint32, bool) {
	for idx, cp := range input {
		if isFullwidthHalfwidth(cp) {
			return idx, cp, true
		}
	}
	return 0, 0, false
}

// rdFirstCombiningStack returns the first base position (a non-Extend codepoint)
// immediately followed by at least minStack consecutive Extend codepoints,
// together with minStack, as (basePos, stackLen, found).
func rdFirstCombiningStack(input []uint32, minStack int) (int, int, bool) {
	for idx, cp := range input {
		if rdIsGraphemeExtend(cp) {
			continue
		}
		following := 0
		for offset := idx + 1; offset < len(input) && following < minStack; offset++ {
			if !rdIsGraphemeExtend(input[offset]) {
				break
			}
			following++
		}
		if following == minStack {
			return idx, minStack, true
		}
	}
	return 0, 0, false
}

// ─────────────────────────────────────────────────────────────────────
// §6 Top-level detection
// ─────────────────────────────────────────────────────────────────────

// rendererDivergenceDetect is the RendererDivergence detection function.
func rendererDivergenceDetect(input []uint32) rdVerdict {
	vsCount := rdCountVs(input)
	combiningCount := rdCountCombining(input)
	fullwidthCount := rdCountFullwidth(input)
	hasZwj := rdInputHasZwj(input)
	ltrCount := rdCountStrongLtr(input)
	rtlCount := rdCountStrongRtl(input)

	inputCopy := make([]uint32, len(input))
	copy(inputCopy, input)

	classification := rdClassify(input, hasZwj, ltrCount, rtlCount)

	return rdVerdict{
		input:          inputCopy,
		classify:       classification,
		vsCount:        vsCount,
		combiningCount: combiningCount,
		fullwidthCount: fullwidthCount,
		hasZwj:         hasZwj,
		strongLtrCount: ltrCount,
		strongRtlCount: rtlCount,
	}
}

// rdClassify walks the five variance triggers in priority order, mirroring the
// Rust reference's nested match. Every arm is explicit; the terminal else is the
// documented Clear verdict, not a catch-all default.
func rdClassify(input []uint32, hasZwj bool, ltrCount, rtlCount int) rdClassification {
	// Priority 1: combining-mark stack overflow (Zalgo).
	if basePos, stackLen, ok := rdFirstCombiningStack(input, rdMinCombiningStack); ok {
		return rdClassification{
			sub:       rdSubThreat{tag: "CombiningStackOverflow", basePos: basePos, stackLen: stackLen},
			positions: []int{basePos},
			decoded:   []uint8{},
		}
	}

	// Priority 2: any variation selector triggers presentation variance.
	if pos, cp, ok := rdFirstVsPos(input); ok {
		return rdClassification{
			sub:       rdSubThreat{tag: "VariationSelectorVariance", firstVsPos: pos, firstVsCp: cp},
			positions: []int{pos},
			decoded:   []uint8{},
		}
	}

	// Priority 3: ZWJ-containing input not in the registered RGI set.
	if hasZwj && !ezwjIsRegisteredZwjSequence(input) {
		if pos, ok := rdFirstZwjPos(input); ok {
			return rdClassification{
				sub:       rdSubThreat{tag: "UnregisteredZwjVariance", firstZwjPos: pos},
				positions: []int{pos},
				decoded:   []uint8{},
			}
		}
		return rdClassification{clear: true, positions: []int{}, decoded: []uint8{}}
	}

	// Priority 4: fullwidth/halfwidth.
	if pos, cp, ok := rdFirstFullwidthPos(input); ok {
		return rdClassification{
			sub:       rdSubThreat{tag: "FullwidthVariance", firstFwPos: pos, firstFwCp: cp},
			positions: []int{pos},
			decoded:   []uint8{},
		}
	}

	// Priority 5: mixed direction.
	if ltrCount > 0 && rtlCount > 0 {
		return rdClassification{
			sub:       rdSubThreat{tag: "MixedDirectionVariance", ltrCount: ltrCount, rtlCount: rtlCount},
			positions: []int{},
			decoded:   []uint8{},
		}
	}

	// No documented variance trigger fired: rendering is stable.
	return rdClassification{clear: true, positions: []int{}, decoded: []uint8{}}
}

File: ports/haskell/src/Unicode/Security/Display/RendererDivergence.hs
{-|
Module      : Unicode.Security.Display.RendererDivergence
Description : Detection of codepoint shapes that render differently across stacks.

Haskell port of @Unicode.Security.Display.RendererDivergence@ from unicode-lean,
transliterated byte-faithfully from the verified Rust reference
implementation (the display-layer detector, reason-code letter @D@).

Threat model. An adversary crafts content that renders one way in the auditor's
renderer (a benign glyph or an empty span) and a different way in the consumer's
renderer (a misleading glyph, a wider glyph, or a different sequence). This is
the "fingerprint stability" family — clear inputs render the same across the
renderer cohort the Standard documents as stable.

What the detector draws. A heuristic three-value split, surfaced through the
universal clear/hazard carrier: an input is 'Clear' when none of the documented
variance triggers fire, and otherwise is classified by the first trigger in
priority order — a combining-mark stack overflow, a variation selector, an
unregistered ZWJ shape, a fullwidth/halfwidth codepoint, or mixed direction. It
reuses the port's own tables (the VariationSelectorPayload detector's
'Unicode.Security.CodepointPredicates.isVariationSelector', the grapheme
'Unicode.Segmentation.Grapheme.lookupGCB' @Extend@ class, the EmojiZwjIntegrity
'Unicode.Security.Identity.EmojiZwjIntegrity.isRegisteredZwjSequence' RGI
registry, and the RtlInjection detector's 'Unicode.Security.CodepointPredicates.isStrongLtr'
\/ 'Unicode.Security.CodepointPredicates.isStrongRtl' strong-bidi classes), never a host
rendering or shaping library.

Sub-threats (priority order).

  1. 'CombiningStackOverflow'    — Zalgo-like combining-mark stack @>= 4@ on a base.
  2. 'VariationSelectorVariance' — any variation selector present.
  3. 'UnregisteredZwjVariance'   — a ZWJ-containing input not in the RGI ZWJ set.
  4. 'FullwidthVariance'         — a fullwidth/halfwidth codepoint present.
  5. 'MixedDirectionVariance'    — both strong-LTR and strong-RTL codepoints.
-}
module Unicode.Security.Display.RendererDivergence
  ( minCombiningStack
  , zwj
  , SubThreat
      ( CombiningStackOverflow, VariationSelectorVariance, UnregisteredZwjVariance
      , FullwidthVariance, MixedDirectionVariance
      )
  , subThreatTag
  , Classification (Clear, Hazard)
  , classificationIsClear
  , classificationTag
  , classificationPositions
  , Verdict
      ( Verdict, verdictInput, verdictClassify, verdictVsCount
      , verdictCombiningCount, verdictFullwidthCount, verdictHasZwj
      , verdictStrongLtrCount, verdictStrongRtlCount
      )
  , isVariationSelector
  , isZwj
  , isFullwidthHalfwidth
  , isGraphemeExtend
  , detect
  , reasonCode
  ) where

import Data.Maybe (listToMaybe)

import Unicode.Segmentation.Grapheme (lookupGCB)
import Unicode.Segmentation.GraphemeTables (GCB (Extend))
import Unicode.Security.Identity.EmojiZwjIntegrity (isRegisteredZwjSequence)
import qualified Unicode.Security.CodepointPredicates as Predicates

-- ─────────────────────────────────────────────────────────────────────
-- §1 Constants
-- ─────────────────────────────────────────────────────────────────────

-- | The combining-mark stack depth (on a single base) at or beyond which the
-- input is treated as a Zalgo-style rendering-variance hazard.
minCombiningStack :: Int
minCombiningStack = 4

-- | The ZERO WIDTH JOINER codepoint.
zwj :: Int
zwj = 0x200D

-- ─────────────────────────────────────────────────────────────────────
-- §2 Types
-- ─────────────────────────────────────────────────────────────────────

-- | Sub-threat enumeration for RendererDivergence, in priority order. Each
-- payload carries the position and codepoint values the conformance harness's
-- attribution column reads back, mirroring the Rust @SubThreat@ variants
-- field-for-field.
data SubThreat
  = -- | A combining-mark stack of @stackLen@ marks on the base at @basePos@.
    -- Fields: the base position, then the stack depth tested
    -- (@>= minCombiningStack@).
    CombiningStackOverflow Int Int
  | -- | A variation selector present in the input. Fields: the position of the
    -- first variation selector, then its codepoint.
    VariationSelectorVariance Int Int
  | -- | A ZWJ-containing input not present in the registered RGI ZWJ set; the
    -- 'Int' is the position of the first ZWJ.
    UnregisteredZwjVariance Int
  | -- | A fullwidth/halfwidth codepoint present. Fields: the position of the
    -- first fullwidth/halfwidth codepoint, then that codepoint.
    FullwidthVariance Int Int
  | -- | Both strong-LTR and strong-RTL codepoints in one input. Fields: the
    -- strong-LTR count, then the strong-RTL count.
    MixedDirectionVariance Int Int
  deriving stock (Eq, Show)

-- | Fixture-row tag string for this sub-threat (matches @SubThreat.tag@).
subThreatTag :: SubThreat -> String
subThreatTag (CombiningStackOverflow _basePos _stackLen) = "CombiningStackOverflow"
subThreatTag (VariationSelectorVariance _vsPos _vsCp)     = "VariationSelectorVariance"
subThreatTag (UnregisteredZwjVariance _zwjPos)            = "UnregisteredZwjVariance"
subThreatTag (FullwidthVariance _fwPos _fwCp)             = "FullwidthVariance"
subThreatTag (MixedDirectionVariance _ltrCount _rtlCount) = "MixedDirectionVariance"

-- | Top-level classification for RendererDivergence (stable = 'Clear').
data Classification
  = -- | Rendering is consistent across the documented renderer cohort.
    Clear
  | -- | A documented variance mode fired: the sub-threat, the implicated
    -- codepoint positions, and the decoded-byte projection (always @[]@ here;
    -- kept for shape parity with the Lean @Classification.hazard@).
    Hazard SubThreat [Int] [Int]
  deriving stock (Eq, Show)

-- | True iff the classification is 'Clear' (i.e. stable).
classificationIsClear :: Classification -> Bool
classificationIsClear Clear               = True
classificationIsClear (Hazard _sub _p _d) = False

-- | Human-facing tag for a hazard, or 'Nothing' when clear.
classificationTag :: Classification -> Maybe String
classificationTag Clear                  = Nothing
classificationTag (Hazard sub _pos _dec) = Just (subThreatTag sub)

-- | Implicated positions ('[]' when clear).
classificationPositions :: Classification -> [Int]
classificationPositions Clear                      = []
classificationPositions (Hazard _sub positions _d) = positions

-- | Verdict — the structured output of 'detect' (mirrors the Lean @Verdict@).
data Verdict = Verdict
  { verdictInput           :: ![Int]
    -- ^ The scanned input codepoints.
  , verdictClassify        :: !Classification
    -- ^ The classification verdict.
  , verdictVsCount         :: !Int
    -- ^ Count of variation selectors.
  , verdictCombiningCount  :: !Int
    -- ^ Count of combining (grapheme @Extend@) marks.
  , verdictFullwidthCount  :: !Int
    -- ^ Count of fullwidth/halfwidth codepoints.
  , verdictHasZwj          :: !Bool
    -- ^ Whether the input contains any ZWJ.
  , verdictStrongLtrCount  :: !Int
    -- ^ Count of strong-LTR codepoints.
  , verdictStrongRtlCount  :: !Int
    -- ^ Count of strong-RTL codepoints.
  }
  deriving stock (Eq, Show)

-- ─────────────────────────────────────────────────────────────────────
-- §3 Core predicates
-- ─────────────────────────────────────────────────────────────────────

-- | True iff @cp@ is a variation selector — reuses the VariationSelectorPayload
-- detector's predicate (ranges @FE00@..@FE0F@, @E0100@..@E01EF@, @180B@..@180D@).
isVariationSelector :: Int -> Bool
isVariationSelector = Predicates.isVariationSelector

-- | True iff @cp@ is the ZWJ codepoint.
isZwj :: Int -> Bool
isZwj cp = cp == zwj

-- | True iff @cp@ is in the Halfwidth/Fullwidth Forms range @FF01@..@FFEF@.
isFullwidthHalfwidth :: Int -> Bool
isFullwidthHalfwidth cp = 0xFF01 <= cp && cp <= 0xFFEF

-- | True iff @cp@ has @Grapheme_Cluster_Break = Extend@ — a combining mark that
-- stacks onto the preceding base. Reuses the port's grapheme-segmentation table
-- via 'lookupGCB'; the RendererDivergence detector uses it to measure
-- Zalgo-style combining-mark stacks.
isGraphemeExtend :: Int -> Bool
isGraphemeExtend cp = lookupGCB cp == Extend

-- ─────────────────────────────────────────────────────────────────────
-- §4 Sub-detectors
-- ─────────────────────────────────────────────────────────────────────

countVs :: [Int] -> Int
countVs input = length (filter isVariationSelector input)

countCombining :: [Int] -> Int
countCombining input = length (filter isGraphemeExtend input)

countFullwidth :: [Int] -> Int
countFullwidth input = length (filter isFullwidthHalfwidth input)

inputHasZwj :: [Int] -> Bool
inputHasZwj input = any isZwj input

countStrongLtr :: [Int] -> Int
countStrongLtr input = length (filter Predicates.isStrongLtr input)

countStrongRtl :: [Int] -> Int
countStrongRtl input = length (filter Predicates.isStrongRtl input)

-- | Position and codepoint of the first variation selector.
firstVsPos :: [Int] -> Maybe (Int, Int)
firstVsPos input =
  listToMaybe [ (idx, cp) | (idx, cp) <- zip [0 ..] input, isVariationSelector cp ]

-- | Position of the first ZWJ.
firstZwjPos :: [Int] -> Maybe Int
firstZwjPos input =
  listToMaybe [ idx | (idx, cp) <- zip [0 ..] input, isZwj cp ]

-- | Position and codepoint of the first fullwidth/halfwidth codepoint.
firstFullwidthPos :: [Int] -> Maybe (Int, Int)
firstFullwidthPos input =
  listToMaybe [ (idx, cp) | (idx, cp) <- zip [0 ..] input, isFullwidthHalfwidth cp ]

-- | The first base position (a non-@Extend@ codepoint) immediately followed by
-- @minStack@ consecutive @Extend@ codepoints. Returns @(basePos, minStack)@ on
-- hit.
firstCombiningStack :: [Int] -> Int -> Maybe (Int, Int)
firstCombiningStack input minStack = go 0 input
  where
    go :: Int -> [Int] -> Maybe (Int, Int)
    go _idx []          = Nothing
    go idx (cp : rest)
      | not (isGraphemeExtend cp) =
          let following = take minStack rest
          in if length following == minStack && all isGraphemeExtend following
               then Just (idx, minStack)
               else go (idx + 1) rest
      | otherwise = go (idx + 1) rest

-- ─────────────────────────────────────────────────────────────────────
-- §5 Top-level detection
-- ─────────────────────────────────────────────────────────────────────

-- | The RendererDivergence detection function. Priority ladder mirrors the
-- verified Rust reference exactly; see the module header for the sub-threat
-- inventory.
detect :: [Int] -> Verdict
detect input =
  Verdict
    { verdictInput          = input
    , verdictClassify       = classification
    , verdictVsCount        = vsCount
    , verdictCombiningCount = combiningCount
    , verdictFullwidthCount = fullwidthCount
    , verdictHasZwj         = hasZwj
    , verdictStrongLtrCount = ltrCount
    , verdictStrongRtlCount = rtlCount
    }
  where
    vsCount        = countVs input
    combiningCount = countCombining input
    fullwidthCount = countFullwidth input
    hasZwj         = inputHasZwj input
    ltrCount       = countStrongLtr input
    rtlCount       = countStrongRtl input

    classification =
      -- Priority 1: combining-mark stack overflow (Zalgo).
      case firstCombiningStack input minCombiningStack of
        Just (basePos, stackLen) ->
          Hazard (CombiningStackOverflow basePos stackLen) [basePos] []
        Nothing ->
          -- Priority 2: any variation selector triggers presentation variance.
          case firstVsPos input of
            Just (pos, cp) ->
              Hazard (VariationSelectorVariance pos cp) [pos] []
            Nothing ->
              -- Priority 3: a ZWJ-containing input not in the registered RGI set.
              if hasZwj && not (isRegisteredZwjSequence input)
                then case firstZwjPos input of
                  Just pos -> Hazard (UnregisteredZwjVariance pos) [pos] []
                  Nothing  -> Clear
                else
                  -- Priority 4: fullwidth/halfwidth.
                  case firstFullwidthPos input of
                    Just (pos, cp) ->
                      Hazard (FullwidthVariance pos cp) [pos] []
                    Nothing ->
                      -- Priority 5: mixed direction.
                      if ltrCount > 0 && rtlCount > 0
                        then Hazard (MixedDirectionVariance ltrCount rtlCount) [] []
                        else Clear

-- | Fully-qualified reason code for a fired sub-threat, of the shape
-- @unicode.security.D.renderer-divergence.\<subThreatTag\>@.
reasonCode :: SubThreat -> String
reasonCode sub = "unicode.security.D.renderer-divergence." ++ subThreatTag sub
